File: Day226/extraCharactersInAString.go
package main

import "fmt"

func minExtraChar(s string, dictionary []string) int {
	memo := make(map[int]int)
	answer := dfs(s, dictionary, memo)
	return answer
}

func dfs(str string, dict []string, memo map[int]int) int {
	if len(str) == 0 {
		return 0
	}
	if val, ok := memo[len(str)]; ok {
		return val
	}
	min := len(str)
	for _, word := range dict {
		if str == word {
			min = 0
			memo[len(str)] = min
			return min
		} else if len(str) >= len(word) && str[:len(word)] == word {
			answer := dfs(str[len(word):], dict, memo)
			min = minInt(min, answer)
		}
	}
	answer := 1 + dfs(str[1:], dict, memo)
	min = minInt(min, answer)
	memo[len(str)] = min
	return min
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	s := "leetcode"
	dictionary := []string{"leet", "code"}
	fmt.Println(minExtraChar(s, dictionary)) // Output: 0
}

File: Day226/extraCharactersInAString.java
public int minExtraChar(String s, String[] dictionary) {
        Map<Integer, Integer> map = new HashMap<>();
        int answer = dfs(s, dictionary, map);
        return answer;
    }
    public int dfs(String str, String[] dict, Map<Integer, Integer> map) {
        if (str.length() == 0) {
            return 0;
        }
        if (map. containsKey(str.length())) {
            return map.get(str.length());
        }
        int min = str.length();
        for (String words: dict) {
            if (str.equals(words)) {
                min = 0;
                map.put(str.length(), min);
                return min;
            } else if (str.startsWith(words)) {
                int answer = dfs(str.substring(words.length()), dict, map);
                min = Math.min(min, answer);
            }
        }
        int answer = 1 + dfs(str.substring(1), dict, map);
        min = Math.min(min, answer);
        map.put(str.length(), min);
        return min;
    }